tests/spec/test_json_schema_spec.py

```diff
@@ -1,6 +1,5 @@
 import json
 import os
-from curses import meta
```

Comment:
Any thoughts on switching to another? I just picked black 'cuz. No real reason. Python ecosystem isn't my strong suit.

Reply:
Big question, I will write up something serious and get back to you.

Reply:
Here's a summary of python tools that tell you something's wrong with your code and maybe try to fix it:

Category | Tool
-- | --
Code formatters | black, autopep8, isort, docformatter, pydocstringformatter
Type checkers | pyright, mypy, pytpe, pyre-check
Error/smell checkers | pylint, pyflakes (+autoflake)
Multi-tools | flake8, ruff
Other checkers | mccabe (complexity), bandit (security), yesqa (remove type ignores), semgrep (generic platform)


There are other feedback/fix mechanisms that do other things, but if you're just wanting something that removes unused imports your options appear to be:

* [autoflake](https://pypi.org/project/autoflake/) - all this does is remove unused imports and variables (using pyflakes
* [ruff](https://pypi.org/project/ruff/) is an all-in-one tool and supports [removing unused imports](https://beta.ruff.rs/docs/rules/unused-import/)
* [pylint](https://pypi.org/project/pylint/) can [warn about it](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/unused-import.html)
* [pyflakes](https://pypi.org/project/pyflakes/) can warn about it. also [flake8](https://www.flake8rules.com/rules/F401.html) (using pyflakes)

`autoflake8` is the most targeted solution. If you wanted a single tool to rule them all then `ruff` looks pretty great (and is rust fast!) but is in beta. `pyflakes` and `pylint` do a a handful of things and are older+slower.


In terms of how to integrate a tool into the dev workflow, you could set up:

* editor integration
* pre-commit hooks
* documented commands
* CI/CD checks as a backstop to the other faster feedback mechanisms

Reply:
Thanks, this is helpful! I've been seeing ruff popping up, but I auto-discount it because it's written in Rust and all things Rust get a premium on the interwebs. I'll take a look at it; it would be nice to just have one thing.

Reply:
Opened https://github.com/recap-build/recap/issues/359